### PR TITLE
Include today's district heating share for myopic optimisation

### PR DESCRIPTION
**Pull request by lisazeyen**
- include today's district heating share based on Hotmaps data
- add option to increase today's district heating share during planning horizion up to maximum level defined in ` central_fraction`
- should be pulled after #136

### CODE REVIEW

File: Snakefile [hunk 1]
```diff
@@ -323,6 +323,7 @@ rule prepare_sector_network:
         transport_name='resources/transport_data.csv',
         traffic_data_KFZ = "data/emobility/KFZ__count",
         traffic_data_Pkw = "data/emobility/Pkw__count",
+        dh_share = "data/district_heat_share.csv",
```
[the remark this answers is not in the document]
Comment (lisazeyen) [reply Snakefile:342]: The data comes from the JRC-IDEES for most countries and for the missing one's from [euroheat](https://www.euroheat.org/knowledge-hub/country-profiles). Sources are now added to the` doc/data.csv`

File: scripts/prepare_sector_network.py [hunk 2]
```diff
@@ -1215,15 +1214,22 @@ def add_heat(n, costs):
...
                 factor = 1 - urban_fraction[nodes[name]]
-            elif "urban" in name:
-                factor = urban_fraction[nodes[name]]
+            elif "urban central" in name:
+                factor = dist_fraction[nodes[name]]
+            elif "urban decentral" in name:
+                factor = urban_fraction[nodes[name]] - \
+                    dist_fraction[nodes[name]]
+            else:
+                factor = None
```
[the remark this answers is not in the document]
Comment (lisazeyen) [reply scripts/prepare_sector_network.py:1383]: Yes, absolutely right, an Error is raised now.

File: config.default.yaml [hunk 3]
```diff
@@ -143,6 +143,12 @@ existing_capacities:
 sector:
   central: true
   central_fraction: 0.6
+  district_heating_increase: true
+  dh_strength:
```
[the remark this answers is not in the document]
Comment (lisazeyen) [reply config.default.yaml:147]: The names are changed now to be a bit clearer about their meaning.

File: scripts/prepare_sector_network.py [hunk 4]
```diff
@@ -1503,24 +1509,55 @@ def create_nodes_for_heat_sector():
...
     # urban can be split into district heating (central) and individual heating (decentral)
-    
+
+    ct_urban = pop_layout.urban.groupby(pop_layout["ct"]).sum()
+    pop_layout["urban_ct_fraction"] = pop_layout["urban"] / \
+            pop_layout["ct"].map(ct_urban.get)
+    # todays district heating share per country
+    dist_heat_share_ct = pd.read_csv(snakemake.input.dh_share, index_col=0,
+                                  usecols=[0,1]).dropna()/100
+    dist_heat_share = pop_layout.ct.map(dist_heat_share_ct["district heating share"])
```
[the remark this answers is not in the document]
Comment (lisazeyen) [reply scripts/prepare_sector_network.py:1676]: Is included now!

File: scripts/prepare_sector_network.py [hunk 5]
```diff
@@ -1503,24 +1509,55 @@ def create_nodes_for_heat_sector():
...
+
+
     sectors = ["residential", "services"]
-    
+
     nodes = {}
+    urban_fraction = pop_layout["urban"] / \
+                     (pop_layout[["urban", "rural"]].sum(axis=1))
```
[the remark this answers is not in the document]
Comment (lisazeyen) [reply scripts/prepare_sector_network.py:1682]: somehow `pop_layout.total` is not the same as `pop_layout[["urban", "rural"]].sum(axis=1)`. For now the old version summing up urban and rural population densities is kept.

File: scripts/prepare_sector_network.py [hunk 6]
```diff
@@ -636,6 +635,9 @@ def prepare_data(n):
 
     nodal_energy_totals = energy_totals.loc[pop_layout.ct].fillna(0.)
     nodal_energy_totals.index = pop_layout.index
+    # district heat share not weighted by population
+    district_heat_share = round(nodal_energy_totals["district heat share"],
+                                ndigits=2)
```
Comment (fneum) [inline scripts/prepare_sector_network.py:640]:
```suggestion
    district_heat_share = nodal_energy_totals["district heat share"].round(2)
```

File: scripts/prepare_sector_network.py [hunk 7]
```diff
@@ -2191,17 +2218,18 @@ def limit_individual_line_extension(n, maxext):
...
             'prepare_sector_network',
             simpl='',
-            clusters="45",
+            opts="",
+            clusters="37",
             lv=1.0,
-            opts='',
             sector_opts='Co2L0-168H-T-H-B-I-solar3-dist1',
+            planning_horizons="2020",
             planning_horizons="2030",
```
Comment (fneum) [inline scripts/prepare_sector_network.py:2233]:
```suggestion
```

File: scripts/prepare_sector_network.py [hunk 8]
```diff
@@ -1661,23 +1669,42 @@ def create_nodes_for_heat_sector():
...
+    progress = get(options["district_heating"]["potential"], investment_year)
+    dist_fraction_node += diff * progress
+    print("************************************")
+    print(
+        "the current DH share compared to the maximum possible is increased \
+           \n by a progress factor of ",
+        progress,
+        "resulting DH share: ",
+        dist_fraction_node)
+    print("**********************************")
```
Comment (fneum) [inline scripts/prepare_sector_network.py:1705]:
```suggestion
    logger.info(
        "The current district heating share compared to the maximum",
        "possible is increased by a progress factor of ",
        progress,
        "resulting in a district heating share of ",
        dist_fraction_node
    )
```

**Review by fneum — APPROVED** [review]
Code looks good. Will do a small comparison tomorrow.

File: scripts/prepare_sector_network.py [hunk 9]
```diff
@@ -2254,10 +2282,10 @@ def limit_individual_line_extension(n, maxext):
         if o == "biomasstransport":
             options["biomass_transport"] = True
 
-    nodal_energy_totals, heat_demand, ashp_cop, gshp_cop, solar_thermal, transport, avail_profile, dsm_profile, nodal_transport_data = prepare_data(n)
+    nodal_energy_totals, heat_demand, ashp_cop, gshp_cop, solar_thermal, transport, avail_profile, dsm_profile, nodal_transport_data, district_heat_share = prepare_data(n)
 
     if "nodistrict" in opts:
-        options["central"] = False
+        options["district_heating"]["progress"] = 0.0
```
[the remark this answers is not in the document]
Comment (lisazeyen) [reply scripts/prepare_sector_network.py:2286]: Yes, thanks for spotting that.

**Pull request merged**